### PR TITLE
113 add markdown to descriptions

### LINKED ISSUE #113: Add markdown to error/warning descriptions
Update the descriptions in the phase_validations.py to contain markdown instead of just plain text.  Based on the FIG descriptions at https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/#4

Use these to generate the csv that is stored in https://github.com/cfpb/sbl-content/blob/main/fig-files/validation-spec/2024-validations.csv?plain=1

Update pytests

### PR DESCRIPTION
**Pull request by jcadam14**
Closes #113 

- Updated phase_validations to include markdown based on wagtail descriptions https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/#4
- Updated pytest to be much simpler as it now just needs to do a direct description comparison instead of reformatting due to csv formatting.

Note, currently the pytest is pointing to the raw csv on the branch for https://github.com/cfpb/sbl-content/issues/21.  Once those changes are merged in, the pytest can be pointed back to main.  So the approval of this PR can wait until that other issue is merged and then I'll reupdate the pytests for final approval.

### CODE REVIEW

File: src/regtech_data_validator/checks.py [hunk 1]
```diff
@@ -11,8 +11,8 @@
 
 
 class Severity(StrEnum):
-    ERROR = 'error'
-    WARNING = 'warning'
+    ERROR = 'Error'
+    WARNING = 'Warning'
```
[the remark this answers is not in the document]
Comment (jcadam14) [reply src/regtech_data_validator/checks.py:15]: The CSV had Error and Warning, and I figured if the frontend is displaying a column or section with result.severity having 'error' or 'warning' would look odd.  All caps or at least the first being capitalized draws more attention and doesn't look so weird if the severity is display on it's own

[the remark this answers is not in the document]
Comment (hkeeler) [reply src/regtech_data_validator/checks.py:15]: 👍

File: src/regtech_data_validator/phase_validations.py [hunk 2]
```diff
@@ -638,12 +635,15 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
...
-                        " fees'\n- 'Initial annual charges'"
+                        "When 'action taken' equals 3 (denied), 4 (withdrawn by applicant), or 5 (incomplete), the following fields must all equal 999 (not applicable):\n"
+                        "* 'Interest rate type'\n"
+                        "* 'MCA/sales-based: additional cost for merchant cash advances or other sales-based financing: NA flag'\n"
+                        "* 'Prepayment penalty could be imposed'\n"
+                        "* 'Prepayment penalty exists'\n\n"
+                        "And the following fields must all be blank:\n\n"
+                        "* 'Total origination charges'\n"
+                        "* 'Amount of total broker fees'\n"
+                        "* 'Initial annual charges'"
```
[the remark this answers is not in the document]
Comment (jcadam14) [reply src/regtech_data_validator/phase_validations.py:646]: Ah cool was not familiar with dedent.  Just read the api and yeah that would make this code prettier.  I'll give that a go.  Will possibly need to redo the csv depending on how it spits out stuff but it should match.

File: src/regtech_data_validator/phase_validations.py [hunk 3]
```diff
@@ -182,11 +184,11 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
...
-                    description=(
-                        "When 'credit product' does not equal 977 (other), 'free-form"
-                        " text field for other credit products' must be blank."
-                        " When 'credit product' equals 977, 'free-form text field "
-                        "for other credit products' must not be blank."
+                    description=dedent(
+                        """\
+                        * When 'credit product' does **not** equal 977 (other), 'free-form text field for other credit products' must be blank.
+                        * When 'credit product' equals 977, 'free-form text field for other credit products' must **not** be blank.
+                    """
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:191]: Did `black` format it this way, with the offset `"""`s?

File: src/regtech_data_validator/phase_validations.py [hunk 4]
```diff
@@ -504,9 +504,9 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
                     id="W2006",
                     name="credit_purpose_ff.multi_invalid_number_of_values",
                     description=(
-                        "'Credit purpose' and 'free-form text field for other credit "
-                        "purpose' combined should not contain more than three values. "
-                        "Code 977 (other), within 'credit purpose', does not count "
+                        "* 'Credit purpose' and 'free-form text field for other credit "
+                        "purpose' combined should **not** contain more than three values. "
+                        "Code 977 (other), within 'credit purpose', does **not** count "
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:509]: This one is still the non-`dedent` flavor.

File: src/regtech_data_validator/phase_validations.py [hunk 5]
```diff
@@ -919,14 +933,13 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
...
-                        "period > 12 months, adjustable interest), 4 (initial rate "
-                        "period > 12 months, fixed interest), 5 (initial rate period "
-                        "<= 12 months, adjustable interest), or 6 (initial rate period "
-                        "<= 12 months, fixed interest), 'initial rate period' must "
-                        "be blank. When 'interest rate type' equals 3, 4, 5, or 6, "
-                        "'initial rate period' must not be blank."
+                    description=dedent(
+                        """\
+                        * When 'interest rate type' does **not** equal 3 (initial rate period > 12 months, adjustable interest), \
+                        4 (initial rate period > 12 months, fixed interest), 5 (initial rate period <= 12 months, adjustable interest), \
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:939]: It _should_ be valid Markdown without the trailing `\`. Is that needed to keep in sync with the CSV?

**Review by hkeeler — COMMENTED** [review]
@jcadam14, how are you feeling about all this? Having the multiline strings seems like a bit of an improvement, but it does look a little odd still, and it seems like it's had implications for the CSV file (https://github.com/cfpb/sbl-content/pull/22#pullrequestreview-1961939542). Do you think we should keep it, or revert back to what you had before?

Looks like there are still quite a few descriptions in the old form. If we decide to keep the multiline, we should probably make 'em all like that.

File: src/regtech_data_validator/phase_validations.py (hunk 3 above)
Comment (jcadam14) [reply src/regtech_data_validator/phase_validations.py:191]: Yup

File: src/regtech_data_validator/phase_validations.py (hunk 4 above)
Comment (jcadam14) [reply src/regtech_data_validator/phase_validations.py:509]: I only used dedent on descriptions with actual multiple lines, the \n's.  This is a single line description, but split into multiple strings for readability.

File: src/regtech_data_validator/phase_validations.py (hunk 5 above)
Comment (jcadam14) [reply src/regtech_data_validator/phase_validations.py:939]: The \ are there to break the string up in the python file so you don't get a single line that looks like "* When 'interest rate type' does **not** equal 3 (initial rate period > 12 months, adjustable interest), 4 (initial rate period > 12 months, fixed interest), 5 (initial rate period <= 12 months, adjustable interest)..."  Which technically should break our linting but black has issues with long string reformatting.

**Comment by jcadam14** [comment]
> @jcadam14, how are you feeling about all this? Having the multiline strings seems like a bit of an improvement, but it does look a little odd still, and it seems like it's had implications for the CSV file ([cfpb/sbl-content#22 (review)](https://github.com/cfpb/sbl-content/pull/22#pullrequestreview-1961939542)). Do you think we should keep it, or revert back to what you had before?
> 
> Looks like there are still quite a few descriptions in the old form. If we decide to keep the multiline, we should probably make 'em all like that.

Most descriptions are actually one single line/string, but they're broken up in the python file for readability.  Need to make a distinction there.  The ones that are actually multiline descriptions have been dedented.

Honestly, the python looks cleaner, but the output to the csv and markdown (before being rendered) is uglier.  Depends on if someone using the csv at some point wants a bunch of \n's.

File: src/regtech_data_validator/phase_validations.py [hunk 6]
```diff
@@ -1022,8 +1093,10 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
                     is_greater_than,
                     id="W0441",
                     name="pricing_adj_margin.unreasonable_numeric_value",
-                    description=(
-                        "When present, 'adjustable rate transaction: margin' should generally be greater than 0.1."
+                    description=dedent(
+                        """\
+                        "* When present, 'adjustable rate transaction: margin' should generally be greater than 0.1."
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:1098]:
```suggestion
                        * When present, 'adjustable rate transaction: margin' should generally be greater than 0.1.
```

File: src/regtech_data_validator/phase_validations.py [hunk 7]
```diff
@@ -1038,9 +1111,11 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
                     is_valid_enum,
                     id="E0460",
                     name="pricing_adj_index_name.invalid_enum_value",
-                    description=(
-                        "'Adjustable rate transaction: index name' must equal "
+                    description=dedent(
+                        """\
+                        "* 'Adjustable rate transaction: index name' must equal "
                         "1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 977, or 999."
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:1117]:
```suggestion
                        * 'Adjustable rate transaction: index name' must equal
                        1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 977, or 999.
```

File: src/regtech_data_validator/phase_validations.py [hunk 8]
```diff
@@ -1100,8 +1175,10 @@ def get_phase_1_and_2_validations_for_lei(context: dict[str, str] | None = None)
                     max_value=300,
                     id="E0480",
                     name="pricing_adj_index_name_ff.invalid_text_length",
-                    description=(
-                        "'Adjustable rate transaction: index name: other' must not exceed 300 characters in length."
+                    description=dedent(
+                        """\
+                        "* 'Adjustable rate transaction: index name: other' must **not** exceed 300 characters in length."
```
Comment (hkeeler) [inline src/regtech_data_validator/phase_validations.py:1180]:
```suggestion
                        * 'Adjustable rate transaction: index name: other' must **not** exceed 300 characters in length.
```

File: tests/test_csv_to_code_differences.py [hunk 9]
```diff
@@ -36,7 +13,7 @@ def test_csv_differences(self):
         ]
 
         csv_df = pd.read_csv(
-            "https://raw.githubusercontent.com/cfpb/sbl-content/main/fig-files/validation-spec/2024-validations.csv"
+            "https://raw.githubusercontent.com/cfpb/sbl-content/21-update-descriptions-for-markdown/fig-files/validation-spec/2024-validations.csv"
```
Comment (hkeeler) [inline tests/test_csv_to_code_differences.py:16]: Assuming we'll merge PR https://github.com/cfpb/sbl-content/pull/22 first.

```suggestion
            "https://raw.githubusercontent.com/cfpb/sbl-content/main/fig-files/validation-spec/2024-validations.csv"
```

**Review by hkeeler — COMMENTED** [review]
Looking good. Found a little clump with leftover `"`s. Founds similar artifacts over in PR https://github.com/cfpb/sbl-content/pull/22 too. I think just the one little fixup and we're good. Thanks for slogging through all this.

File: tests/test_csv_to_code_differences.py (hunk 9 above)
Comment (jcadam14) [reply tests/test_csv_to_code_differences.py:16]: Good catch thank you.  After I updated I went through each to check for just this but your eyes start to cross after awhile!

Comment (hkeeler) [reply tests/test_csv_to_code_differences.py:16]: Yep. No doubt. 😄

**Review by hkeeler — APPROVED** [review]

**Pull request merged**